docs/RFCS/20200507_partial_indexes.md

```diff
@@ -0,0 +1,318 @@
+
+The time complexity of this check is `O(P * F)`, where `P` is the number of
+nodes in the predicate expression and `F` is the number of nodes in the filter
+expression.
+
+## Generating Partial Index Scans
+
+We will consider utilizing partial indexes for both unconstrained and
+constrained scans. Therefore, we'll need to modify both the `GenerateIndexScans`
+and `GenerateConstrainedScans` exploration rules (or make new, similar rules).
+
+In addition, we'll need to update exploration rules for zig-zag joins and
+inverted index scans.
+
+We'll remove redundant filters from the expression when generating a scan over a
+partial index. For example:
+
+```sql
+CREATE TABLE products (id INT PRIMARY KEY, price INT, units_sold INT, units_in_stock INT)
+CREATE INDEX idx1 ON products (price) WHERE units_sold > 1000
+
+SELECT * FROM products WHERE price > 20 AND units_sold > 1000 AND units_in_stock > 0
+```
+
+When generating the constrained scan over `idx1`, the `units_sold > 1000` filter
+can be removed from the outer `Select`, such that only the `units_in_stock > 0`
+filter remains.
+
+## Statistics
+
+The statistics builder must take into account the predicate expression, in
+addition to the filter expression, when generating statistics for partial index
+scan. This is because the number of rows examined via a partial index scan is
+dependent on the predicate expression.
+
+For example, consider the following table, indexes, and query:
+
+```sql
+CREATE TABLE products (id INT PRIMARY KEY, price INT, units_sold INT, type TEXT)
+CREATE INDEX idx1 ON t (price) WHERE units_sold > 1000
+CREATE INDEX idx2 ON t (price) WHERE units_sold > 1000 AND type = 'toy'
+
+SELECT COUNT(*) FROM products where units_sold > 1000 AND type = 'toy' AND price > 20
+```
+
+A scan on `idx1` will scan `[/1001 - ]`. A scan on on `idx2` will have the same
+scan, `[/1001 - ]`, but will examine fewer rows—only those where `type = 'toy'`.
+Therefore, the optimizer cannot rely solely on the scan constraints to determine
+the number of rows returned from scanning a partial index. It must also take
+into account the selectivity of the predicate to correctly determine that
+scanning `idx2` is a lower-cost plan than scanning `idx1`.
+
+We can estimate the number of rows returned from scanning a partial index with
+the following formula:
+
+    num_rows = rows_in_table * selectivity(predicate_expression) * selectivity(scan_constraint)
+
+This formula is similar one described by Michael Stonebraker in
+["The Case For Partial Indexes"](https://dsf.berkeley.edu/papers/ERL-M89-17.pdf).
+It has been simplified such that it does not make special considerations for
+columns both in the partial index column set and in the partial index
+predicate. In a series of examples, this proved to have an insignificant effect
+on the resulting estimate.
+
+## Mutation
+
+Partial indexes only index rows that satisfy the partial index's predicate
```

Comment:
One tricky thing to make sure you consider is the effect of a partial index that is currently being added or removed and/or if an existing partial index is having columns added or removed. It should follow the same patterns as regular indexes, but make a special point of carefully testing those kinds of variations.